docs/community/20241112-node-bootstrapping.md

```diff
@@ -0,0 +1,118 @@
+bootstrap providers.
+I would like to read and to follow documentation, guidelines, and specifications
+on the above.
+I would like to offer a choice of node bootstrapping configuration systems to users,
+enable distribution vendors to extend the choices available by adding new
+configuration systems, and ease maintenance and enable extension of currently
+supported systems for the ClusterAPI developer community.
+
+
+**Problem statement / Example issues**
+
+Currently, the kubeadm bootstrap provider implements two systems to configure a node's
+operating system at provisoining time (referred to as "OS provisioning" from here on):
+cloud-init and ignition.
+As there is currently no abstract way to pass OS provisioning configuration to node
+bootstrapping, both OS provisioning system implementations - ignition and cloud-init - must cover
+both generic node bootstrapping as well as OS provisioning system specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS provisioning
+systems since kubeadm's implementations is specific to kubeadm and cannot be re-used.
+MicroK8s currently only supports cloud-init.
+
+Outside of ClusterAPI, both [cloud-init](https://github.com/canonical/cloud-init)
+and [ignition](https://github.com/coreos/ignition) provisioning are widely adopted
+across distributions.
+While cloud-init is used by general-purpose Linux distributions like Ubuntu/Debian,
+SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
+with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
+/ SLE Micro / Kalpa, and Flatcar Container Linux.
+It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
+and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
+share implementations across bootstrap providers.
+
+
+**Compatibility and API guarantees**
+
+ClusterAPI has been stable for multiple years and is in widespread production use.
+Proposals and implementations of this working group must ensure to not exert operational
+risk on existing integrations.
+Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
+expectations set by the current implementation and must continue to support current use
+cases OR must provide means for current use cases to continue functioning for an appropriate
+migration period, as well as define a clear path for migration.
+
+## Scope
+
+1. Propose architectural improvements that abstract OS provisioning configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new OS provisioning
+   features, and simplifying integration of new configuration systems for OS provisioning.
+2. Deliver an example implementation in the kubeadm bootstrap provider.
```

Comment:
```suggestion
2. Deliver an example implementation which works with the kubeadm bootstrap provider.
```

The current phrasing implies that provisioning is embedded inside bootstrap -- which is one problem we are trying to address so I'm not sure we want to prescribe this assumption in advance. We may well decide that we want to keep things as they are, but let's at least discuss things first.

Reply:
Good point; and I believe Fabrizio and you are in alignment with this.